src/class/CgiManager/CgiManager.cpp

```diff
@@ -0,0 +1,225 @@
+	  << "Port=\"" << i.getPort() << "\""
+	  << "PathCGI=\"" << i.getPathCGI() << "\""
+	  << "BasicHTTPRequest=\"" << i.getBasicHTTPRequest() << "\""
+	  << "]";
+	return o;
+}
+
+/*
+** --------------------------------- METHODS ----------------------------------
+*/
+
+void CgiManager::_free_argv_env(char **ch_env, char **argv)
+{
+	if (argv != NULL)
+	{
+		for (int i = 0; argv[i] != NULL; i++)
+		{
+			if (argv[i] != NULL)
+				free(argv[i]);
+			argv[i] = NULL;
+		}
+		free(argv);
+		argv = NULL;
+	}
+	if (ch_env != NULL)
+	{
+		for (int i = 0; ch_env[i] != NULL; i++)
+		{
+			if (ch_env[i] != NULL)
+				free(ch_env[i]);
+			ch_env[i] = NULL;
+		}
+		free(ch_env);
+		ch_env = NULL;
+	}
+}
+
+char **CgiManager::_setEnv(void) const
+{
+	std::map<std::string, std::string> env;
+	env[httpConstants::CONTENT_LENGTH] = _basicHTTPRequest.getBody().length();
+	if (_basicHTTPRequest.getHeaders().find(httpConstants::CONTENT_TYPE_FIELD) != _basicHTTPRequest.getHeaders().end())
+		env[httpConstants::CONTENT_TYPE_FIELD] = _basicHTTPRequest.getHeaders().at(httpConstants::CONTENT_TYPE_FIELD);
+	env[httpConstants::REQUEST_METHOD] = _basicHTTPRequest.getType();
+	env[httpConstants::SCRIPT_FILENAME] = _pathCGI.get();
+	if (_basicHTTPRequest.getQuery().length() > 1)
+		env[httpConstants::QUERY_STRING] = _basicHTTPRequest.getQuery().substr(1);
+	env[httpConstants::SERVER_PROTOCOL] = _basicHTTPRequest.getHTTPVersion();
+	env[httpConstants::SERVER_PORT] = _port;
+	env[httpConstants::SERVER_NAME] = _serverName;
+	env[httpConstants::PATH_INFO] = _basicHTTPRequest.getPath();
+
+	char **ch_env = (char **)calloc(sizeof(char *), env.size() + 1);
+	if (!ch_env)
+		throw CgiManager::CgiManagerException();
+	std::map<std::string, std::string>::const_iterator it = env.begin();
+	for (int i = 0; it != env.end(); it++, i++)
+	{
+		std::string tmp = it->first + "=" + it->second;
+		ch_env[i] = strdup(tmp.c_str());
+	}
+	ch_env[env.size()] = NULL;
+	return (ch_env);
+}
+
+const std::string CgiManager::_createpipe(char **ch_env, char **argv)
+{
+	int _exit_status;
+	if (pipe(_inputPipe) < 0 || pipe(_outputPipe) < 0)
+		throw CgiManager::CgiManagerException();
+	int pid = fork();
+	if (pid < 0)
+		throw CgiManager::CgiManagerException();
+	if (pid == 0)
+	{
+		dup2(_inputPipe[0], STDIN_FILENO);
+		dup2(_outputPipe[1], STDOUT_FILENO);
+		close(_inputPipe[0]);
+		close(_inputPipe[1]);
+		close(_outputPipe[0]);
+		close(_outputPipe[1]);
+		_exit_status = execve(argv[0], argv, ch_env);
+		exit(_exit_status);
+	}
+	else if (pid > 0)
+	{
+		close(_inputPipe[0]);
+		close(_outputPipe[1]);
+		write(_inputPipe[1], _basicHTTPRequest.getBody().c_str(), _basicHTTPRequest.getBody().length());
+		close(_inputPipe[1]);
+
+		char buffer[40000 * 2];
+		size_t bytes_read = 0;
+		std::string output;
+		waitpid(pid, &_exit_status, 0);
+		std::size_t pos;
+		int contentLenght;
+		while ((bytes_read = read(_outputPipe[0], buffer, sizeof(buffer))) > 0)
```

Comment:
again about `read` without `poll`

Reply:
also i think this section will be very similar to reading a request from a client so you can use BasiHTTPRequest again for the parsing and checking the header.

Reply:
I think it is not so similar. Depending on the cgi you usually send at most content-length, content-type and body. With this only BasiHTTPRequest would give error.